### ISSUE TITLE
[R-package] running error with multiclass

### ISSUE THREAD

**Issue by zheguzai100** [Microsoft/LightGBM#716]
`library(lightgbm)`
`wine1 <- read.csv('G:/rwork/wine/whitewines.csv',header = TRUE)`
`indx1 <- sample(1:2,nrow(wine1),prob = c(0.8,0.2),replace = TRUE)`
`train1 <- wine1[indx1==1,]`
`test1 <- wine1[indx1==2,]`
`dtrain1 <- lgb.Dataset(data=as.matrix(train1[,-ncol(train1)]),label=train1$quality-3)`
`dtest1 <- lgb.Dataset.create.valid(dtrain1,data = as.matrix(test1[,-ncol(test1)]),label=test1$quality-3)`
`valids <- list(eval=dtest1,train=dtrain1)`
`params <- list(objective='multiclass',metric='multi_err',num_class=7)`

`lgb1 <- lgb.train(params = params,data = dtrain1,#valids = valids,`
`                  num_leaves = 6,
                  nrounds = 100,early_stopping_rounds = 10,
                  min_data=1,learning_rate=1)`

[LightGBM] [Info] Total Bins 1240
[LightGBM] [Info] Number of data: 3917, number of used features: 11
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=3
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=4
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=3
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=4
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=3
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=5
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=3
Error in init(env) : 
  For early stopping, valids must have at least one element

**Comment by guolinke** [comment]
I am not familiar with R, what is ```#valids = valids``` ?

**Comment by wxchan** [comment]
Isn't this just because `valids` is commented out?

**Comment by Laurae2** [comment]
`#valids` is commented out therefore there is no validation dataset hence LightGBM complains about `early_stopping_rounds` being used without validation dataset.

**Comment by zheguzai100** [comment]
@guolinke  
 sorry for this confussion, last time i run `lgb.train(params = params,data = dtrain1, valids = valids, .....` . it brought error above, so i commented it out as `#valids = valids`, but it still brought the same error.

**Comment by Laurae2** [comment]
@zheguzai100 when you have a LightGBM training error due to wrong parameters, refer to https://github.com/Microsoft/LightGBM/pull/712 for the solution.

https://github.com/Microsoft/LightGBM/blob/master/docs/FAQ.md#r-package

**Comment by zheguzai100** [comment]
@Laurae2 
you mean adding `lgb.unloader(wipe=TRUE)` ?? 
it doesn't work
here's my [data](https://github.com/zheguzai100/myRproj/commit/475bbca5843994019ef9e5a566be78697abb20ef)

**Comment by zheguzai100** [comment]
after removing `early_stopping_rounds` Rscripts runs well.


and then another question surfaced:  
1. by setting `metric='multi_err'`, where is  multi_err's result ? cause i just got info bellow,
which can be  my base to tune my `params` in `lgb.train` ???
`[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=5
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=3
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=4
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=4
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=4
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=5
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=5
[LightGBM] [Info] Trained a tree with leaves=6 and max_depth=4
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Info] No further splits with positive gain, best gain: -inf`
 
2. after `pred1 <- predict(model,as.matrix(test[,xx]),reshape=TRUE)` ,  how could i get the final classification result?  is it apply(pred1,1, function(x)which.max(x)) ??
BTW, what's value meaning for in pred1, possibility to a class??
if so, `table(test1$quality-3==pre1)` 
`FALSE  TRUE` 
 `6702    18` 
unbelievable, is there anything wrong in my Rscript?

**Comment by guolinke** [comment]
@zheguzai100 
1. it is ```multi_error``` . And i think ```early_stopping_rounds ``` can be used if you fix this. 
2. ```min_data=1,learning_rate=1``` will cause the poor accuracy. please use bigger ```min_data``` and smaller ```learning_rate``` 
3. ```pred1``` is the possibility distribution of classes, normalized by softmax.  
4. what is ```pre1``` ? the result of ```apply(pred1,1, function(x)which.max(x))``` ?

**Comment by Laurae2** [comment]
@zheguzai100 any news? You have coding issues (1, snowballing on other hyperparameters) / context of your code missing (4) as @guolinke outlined in the comment before.

**Comment by zheguzai100** [comment]
here is the whole code, [data](https://github.com/zheguzai100/myRproj/commit/475bbca5843994019ef9e5a566be78697abb20ef)
`library(lightgbm)
wine1 <- read.csv('G:/rwork/wine/whitewines.csv',header = TRUE)
indx1 <- sample(1:2,nrow(wine1),prob = c(0.8,0.2),replace = TRUE)
train1 <- wine1[indx1==1,]
test1 <- wine1[indx1==2,]
dtrain1 <- lgb.Dataset(data=as.matrix(train1[,-ncol(train1)]),label=train1$quality-3)
dtest1 <- lgb.Dataset.create.valid(dtrain1,data = as.matrix(test1[,-ncol(test1)]),label=test1$quality-3)
valids <- list(eval=dtest1,train=dtrain1)
params <- list(objective='multiclass',metric='multi_err',num_class=7)
lgb1 <- lgb.train(params = params,data = dtrain1,valids = valids,
                  num_leaves = 6,
                  nrounds = 100,#early_stopping_rounds = 3,#min_data=1,
                  learning_rate=1)
pre1 <- predict(lgb1,data=as.matrix(test1[,-ncol(test1)]),reshape = TRUE)
pre1
pre1.cat <- apply(pre1,1,function(x) {which.max(x)})
#head(test1$quality-3,10);head(pre1.cat,10)
table(test1$quality);table(pre1.cat)
table(test1$quality-3==pre1.cat)
`

**Comment by zheguzai100** [comment]
@guolinke 
sir, info shows on upstair

**Comment by guolinke** [comment]
@Laurae2 can you help for this ?

**Comment by Laurae2** [comment]
1. You need to use the validation metric, and in your case loop through all the range you want to optimize. Then, you take the best validation metric.

2. There is some code available in a closed issue in this repository for hyperparameter tuning for R. Search for "expand.grid" and you should find it.

**Comment by guolinke** [comment]
@zheguzai100 try bigger min_data, and small learning_rate .
